### ISSUE TITLE
CustomRecurrentLayer with DropoutLayer in hid_to_hid fails to compile

### ISSUE THREAD

**Issue by f0k** [Lasagne/Lasagne#429]
From the mailing list: https://groups.google.com/forum/#!topic/lasagne-users/8E1iR04_nK8

The following code fails:
```python
import theano
import lasagne
n_batch, n_steps, n_in = (2, 3, 4)
n_hid = 5
l_in = lasagne.layers.InputLayer((n_batch, n_steps, n_in))
l_in_hid = lasagne.layers.DenseLayer(
    lasagne.layers.InputLayer((None, n_in)), n_hid)
l_hid_hid = lasagne.layers.DropoutLayer(lasagne.layers.DenseLayer(
    lasagne.layers.InputLayer((None, n_hid)), n_hid))
l_rec = lasagne.layers.CustomRecurrentLayer(l_in, l_in_hid, l_hid_hid)
out = lasagne.layers.get_output(l_rec)
fn = theano.function([l_in.input_var], out)
```
The exception is:
```
MissingInputError: ("An input of the graph, used to compute dot(<TensorType(float32, matrix)>, HostFromGpu.0), was not provided and not given a value.Use the Theano flag exception_verbosity='high',for more information on this error.", <TensorType(float32, matrix)>)
```
With exception_verbosity='high', it is:
```
theano.gof.fg.MissingInputError: A variable that is an input to the graph was neither provided as an input to the function nor given a value. A chain of variables leading from this input to an output is [<TensorType(float32, matrix)>, dot.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{mul,no_inplace}.0, Elemwise{true_div,no_inplace}.0, Shape.0, Elemwise{Cast{int32}}.0, GPU_mrg_uniform{CudaNdarrayType(float32, matrix),no_inplace}.0]. This chain may not be unique
Backtrace when the variable is created:
  File "/raid/user/jan/install/Lasagne-git/lasagne/layers/recurrent.py", line 451, in get_output_for
    strict=True)[0]
```
That line is: https://github.com/Lasagne/Lasagne/blob/fe2608c9c675a437dd5bbcb577e6f09817fc57bc/lasagne/layers/recurrent.py#L451

It is the call to `theano.scan`.

The error disappears when disabling the dropout layer (by setting `p=0`). Is this a Theano bug, or is there something special to do when using a symbolic RNG within scan?

**Comment by f0k** [comment]
Ah, there was another important detail: The error also disappears when setting the input layer shape in `l_hid_to_hid` to `n_batch, n_hid` instead of `None, n_hid`. When doing so, the `DropoutLayer` will use that fixed shape to generate the dropout mask instead of the symbolic input shape: https://github.com/Lasagne/Lasagne/blob/fe2608c9c675a437dd5bbcb577e6f09817fc57bc/lasagne/layers/noise.py#L77-L80
What's the variable that `DropoutLayer` will see when used in `CustomRecurrentLayer`? Is there anything odd with it?

**Comment by f0k** [comment]
Somewhat minimal working example in pure Theano:
```python
import theano
import theano.tensor as T
from theano.sandbox.rng_mrg import MRG_RandomStreams as RandomStreams
floatX = theano.config.floatX

inputs = T.matrix()
hid_init = T.vector()
rng = RandomStreams()

# Works:
def step(inp, prev):
	# rng.binomial with constant shape
    return inp + prev * rng.binomial((10,), p=.5, dtype=floatX)
out = theano.scan(step, sequences=[inputs], outputs_info=[hid_init])[0]
theano.function([inputs, hid_init], out)
print "Success!"

# Fails:
def step(inp, prev):
	# rng.binomial with symbolic shape
    return inp + prev * rng.binomial(prev.shape, p=.5, dtype=floatX)
out = theano.scan(step, sequences=[inputs], outputs_info=[hid_init])[0]
theano.function([inputs, hid_init], out)
print "Will never be reached!"
```
The error is the same, it's a theano.gof.fg.MissingInputError.

@nouiz, any idea? Shall we report this as a Theano issue?

**Comment by f0k** [comment]
Reported upstream as https://github.com/Theano/Theano/issues/3437.

**Comment by f0k** [comment]
It's not a Theano bug: `theano.scan` has two return values, the second of which is an updates dictionary. `CustomRecurrentLayer` just discards it, while it should have been caught and passed to `theano.function` (as documented on http://deeplearning.net/software/theano/library/scan.html#using-shared-variables-gibbs-sampling). Even in the cases where compilation doesn't fail with a `MissingInputError`, it will cause each function call to use the very same dropout masks.

One way to fix this would be changing the `get_output()` API to return an update dictionary along with the output expression. Changing the API is something we should have done before Lasagne 0.1, though.
Instead of having `get_output()` return a second output, we could use `**kwargs` to pass an `updates` dictionary that is extended in a layer's `get_output_for()` if needed. Users would have to be made aware of the need for passing an `updates` dictionary (by raising an exception or a warning) if they didn't pass one but should (i.e., the recurrent layers would only need to do this when `scan` returns a non-empty `updates` dictionary).
In any case, since we need `get_output()` to handle updates, we could also use this for the `BatchNormLayer` instead of the `default_updates` trick.

To reiterate, I'd propose changing the recurrent layer's `get_output_for()` to something like:
```python
def get_output_for(self, inputs, **kwargs):
    ...
    output, updates = theano.scan(...)
    if updates:
        if not 'updates' in kwargs:
            raise ValueError("%r requires an `updates` dictionary to be "
                             "passed to `get_output()` or `get_output_for()` for "
                             "recording updates to internal Theano variables. "
                             "This needs to be passed to `theano.function()` "
                             "later on." % self)
        kwargs['updates'].update(updates)  # that one reads great, doesn't it?
    ...
```

**Comment by skaae** [comment]
Nice work investigating this. Is the problem that the random variables need the updates dict to be passed to the theano function? 

In that case I think your solution is good. We can also descripte the problem in the notes section.

**Comment by f0k** [comment]
> Is the problem that the random variables need the updates dict to be passed to the theano function?

Yes, exactly.

**Comment by craffel** [comment]
> To reiterate, I'd propose changing the recurrent layer's get_output_for() to something like:

This seems reasonable to me, and arguably a little cleaner than returning multiple things from `get_output` as most of the time people won't need to use this.  Do you want me to implement this?

**Comment by f0k** [comment]
> Do you want me to implement this?

Yes, please feel free! Maybe also update the RNN example to pass an `updates` OrderedDict, even if it's not actually needed there.

**Comment by craffel** [comment]
Cool, please assign to me.

**Comment by craffel** [comment]
A snag: `lasagne.utils.unroll_scan` doesn't support updates.  We can either modify it to support updates, (@skaae any ideas?) or we can come up with a way to check whether updates are needed.  The simplest way I can think of doing the latter is calling `theano.scan` even in the `unroll_scan=True` case and `if updates and self.unroll_scan`, raise an error, which seems hacky and wrong.  @f0k can you think of any ways to traverse the variables involves in scan's function to see if any of them will require updates?

**Comment by f0k** [comment]
I think if you unroll the scan, you won't need to take care about the updates, the `default_updates` mechanism will work fine. We don't need to care about updates in a simple feed-forward network either. It would only be needed if `fn` (the step function) explicitly returned an update dictionary, but we're not using this feature, are we? It might be good to document this limitation in `unroll_scan`, but it's actually obvious since it doesn't return an `updates` dictionary.

**Comment by craffel** [comment]
Ah I see, ok.  Well, I will add a test for the `unroll_scan` case too to ensure that it works the same when there are updates.

**Comment by skaae** [comment]
I came across this default update hack. 

```
for k, v in rnn_updates.iteritems():
        k.default_update = v
```

Make updates returned by scan  ``default_update``. Would that fail in some cases?

**Comment by f0k** [comment]
> I came across this default update hack. [...] Would that fail in some cases?

I don't know. If it was that easy, I assume `theano.scan` would do it like this and not leave the burden up to the user. @nouiz?

If we can handle this case with default updates, then we don't need an extra interface for the updates at all. We could use this for state-preserving recurrent layers as well (by assigning a default update to the shared variables holding the hidden states).

**Comment by mihaha** [comment]
What's the currently with this Issue? Why isn't it in master yet?
Clearly, it's a bug, just make a bugfix to all recurrent layers)

**Comment by f0k** [comment]
> Why isn't it in master yet?

Because nobody took the time to figure out yet whether the [default_update trick](https://github.com/Lasagne/Lasagne/issues/429#issuecomment-169952568) would cover all cases and not break anything -- if so, this would be the way to go, if not, `get_output()` needs a mechanism to return updates. This issue has probably stalled because it affects so few users. You're very welcome to have a look at this, we could definitely use more contributors! Let me know if you need any advice!
